### ISSUE TITLE
Adding audio / subtitles sync menu in ExoPlayer (adjust audio / subtitles sync functionality) 

### ISSUE THREAD

**Issue by MSTRKKRFT** [jellyfin/jellyfin-androidtv#2479]
### Describe the feature you'd like

Hello!

I noticed that libVLC player allows adjusting audio & subtitles sync, which is great as depending on the file format I have different audio delays. Sometimes the file will play just fine without sync issue, but with Dolby Vision files, for example, I need to add 350ms delay to get audio in sync. 

Because of that (and problems with subtitles in ExoPlayer, as described here : [https://github.com/jellyfin/jellyfin-androidtv/issues/2374#issuecomment-1420531498](url) ), I have to use Kodi as external player if I want to use Jellyfin ATV app, but with that solution watch state tracking is not working, so I would prefer to be able to use ExoPlayer inside jellyfin app. 

Is it possible to add this option in ExoPlayer as well ?

**Comment by tim-vk** [comment]
It's been a while but I'm the guy who added the subtitle sync to the libVLC player and at that time I could not find a plug-and-play method to get subtitle offsets working in exoplayer. See [here](https://github.com/google/ExoPlayer/issues/854).

However!

I saw a very interesting conversation [here](https://github.com/jellyfin/jellyfin-androidtv/pull/2239) where I maybe oversaw a very simple solution for that. So that's something I can maybe take a look at.

I don't have the slightest idea how the audio sync would work: haven't looked at that.

By the way: Your link is dead. Could you check it?

**Comment by MSTRKKRFT** [comment]
First, thanks for your work!
I'm not a dev, so I really can't help you on that… But if you need any end-user testing or something, I would be glad to help you.
Just updated the link ! The problem is integrated subs & Dolby Vision aren't playing great together in jellyfin-exoplayer, but no issue with kodi-exoplayer (maven's build from kodinerds)

**Comment by zerodogg** [comment]
It would be nice if this could get implemented. The lack of audio delay settings in 0.17.x has rendered jellyfin unusable for certain files on my android tv, I’m having to revert to an older version via fdroid.

**Comment by jdumeez** [comment]
with libVLC removed on the android client, it would be great if this topic is picked up again. 
We used the subtitle offset feature all the time

**Comment by itsZeebo** [comment]
I also think it would be amazing if we could have this feature back.
For some older shows it's impossible to find subtitles that are 100% in sync out of the box so you have to manually fix it on the go.

**Comment by M4TH1EU** [comment]
Agree with what has been said, this was a really useful feature and now some movies/shows are hard to watch with unsynced subtitles.

**Comment by mschumacher69** [comment]
Any progress on this?

**Comment by TianZuo555** [comment]
Could you please provide some updates on the subtitle sync?   Just want to see if I could help. Thanks

**Comment by GlobeDotters** [comment]
Hey!

I’d like to perhaps give it a go and try to contribute a subtitle offset control to Android TV, as a first phase toward full sync tools (I know Bazarr can help with sync, but this would be native and controllable via the application UI).

### Approach (ExoPlayer/Media3-compliant)

* Inject a thin **`SubtitleDecoderFactory` wrapper** that shifts cue event times by a runtime `offsetUs`.
* No reflection / no ExoPlayer fork; changes apply live when the user adjusts offset.
* **Scope:** external text tracks (SRT/WebVTT/ASS).
* **Not applicable:** image/PGS or burned-in subtitles.

### UX

* Add **“Subtitle sync”** under ⚙️ with controls:
  `−0.5s   −0.1s   +0.1s   +0.5s   Reset`
* Display the current offset (e.g., `+0.8s`).
* Apply changes immediately during playback.

It could also have a persistence per media file, etc.

If this direction sounds okay, do let me know and I might try to give it a go.

Thanks!

**Comment by austinrbrown01** [comment]
@GlobeDotters That would be awesome! I just ran into this issue myself and checked to see if anything was being actively developed for it. Looks like a highly-requested feature for a while :)

**Comment by GlobeDotters** [comment]
@austinrbrown01 awesome! I think I would just have to hear from one of the devs/maintainers here if they would be ok with it too, and I could give it a shot :)

**Comment by dzek69** [comment]
@GlobeDotters hi, not a maintaniner here, but if you'd make it work, I'd just use your fork until the main version gets this feature. If that would be 5 years later and I'd miss a tons of new features - I don't care. This is the only missing feature I really need.

**Comment by ShadyBoukhary** [comment]
@GlobeDotters any update on this?

**Comment by virtualducttape** [comment]
Agreed this feature is important! @GlobeDotters , I'm not one of the project developers, but your approach seems reasonable to me... especially because doing it as a thin wrapper means it could easily be adjusted on the fly without needing to impact the video playback at all (which hopefully also means it could be used to load/enable/disable the subtitles without effecting the playback), it's incredibly frustrating when any change to subtitles causes the video to rebuffer....

**Comment by s17534** [comment]
@GlobeDotters How is your progress? :)

**Comment by 9acca9** [comment]
Hi. Yes, this is essential. I'm looking forward to this feature too!

Thanks!

**Comment by pulpul-s** [comment]
Now that I have an Android TV device myself, I made a patch that adds this feature. Unfortunately I don't have the time to test my patch properly nor to make sure it complies with all Jellyfin guidelines. However in my testing it was fully working.

Currently it has a "Subtitle offset" option in the subtitle menu if the selected subtitle supports offsetting (for example srt). 
Buttons for +-0.1s +-0.5s offsetting, reset and up and down buttons on the remote sets it +-0.1s (keep pressed for fast setting large offsets).

I'll just leave the patch here and if someone is interested they can either use it as a base, improve upon it or just compile your own apk, like I did. 

[jellyfin-androidtv-subtitle-offset.patch](https://github.com/user-attachments/files/26167807/jellyfin-androidtv-subtitle-offset.patch)

_Please note that I used ChatGPT while making it._

**Comment by ErikDB87** [comment]
@GlobeDotters, have you been able to make progress? 🙏🏻

@pulpul-s, thanks for the patch. I'm trying to make an APK out of it. First time I've tried that, so we'll see how it goes. 

Why not make a PR out of it? If it doesn't follow the guidelines, surely someone will point out what should be done differently ;)

**Comment by ErikDB87** [comment]
I was able to make an APK out of @pulpul-s's patch. It seems to work!

I think it should be made a PR.

**Comment by s17534** [comment]
@ErikDB87 could you describe how to make such APK or point to right direction?

**Comment by ErikDB87** [comment]
> @ErikDB87 could you describe how to make such APK or point to right direction?

Android Studio and some guiding by Claude. :) Here's my APK: <removed>
If needed, I'll open the PR myself later this week.

**Comment by nielsvanvelzen** [comment]
We do not allow for sharing binary files here, as they can be easily used to spread malware.

**Comment by s17534** [comment]
@ErikDB87  That was fast, thank you ☺️

**Comment by ErikDB87** [comment]
> We do not allow for sharing binary files here, as they can be easily used to spread malware.

My apologies.

@s17534, if you were too late to download, I'll try to upload it as a release file what is it called of my fork tomorrow (or Tuesday).

**Comment by s17534** [comment]
@ErikDB87 I'll try do it myself just for education and I hope if this patch is working correctly then such a feature should be incorporated ASAP

**Comment by pulpul-s** [comment]
> [@ErikDB87](https://github.com/ErikDB87) could you describe how to make such APK or point to right direction?

Repository README has a guide on how to build and what are the requirements.
Building the APK should be as simple as:
1. install android sdk
3. clone the repository
4. copy the patch file to cloned repository
5. `git apply jellyfin-androidtv-subtitle-offset-git.patch`
6. `gradlew.bat assembleDebug`


I have included a patch that should work directly with the current master branch of the repository, nothing changed in terms of the code, but I have not tested it.

[jellyfin-androidtv-subtitle-offset-git.patch](https://github.com/user-attachments/files/27954953/jellyfin-androidtv-subtitle-offset-git.patch)

**Comment by ErikDB87** [comment]
> I have included a patch that should work directly with the current repository, nothing changed in terms of the code, but I have not tested it.\n\njellyfin-androidtv-subtitle-offset-git.patch

Why not open a PR? :)

It would be great if this feature made it into the actual stable release.

**Comment by s17534** [comment]
> > [@ErikDB87](https://github.com/ErikDB87) could you describe how to make such APK or point to right direction?
> 
> Repository README has a guide on how to build and what are the requirements. Building the APK should be as simple as:
> 
> 1. install android sdk
> 2. clone the repository
> 3. copy the patch file to cloned repository
> 4. `git apply jellyfin-androidtv-subtitle-offset-git.patch`
> 5. `gradlew.bat assembleDebug`
> 
> I have included a patch that should work directly with the current repository, nothing changed in terms of the code, but I have not tested it.
> 
> [jellyfin-androidtv-subtitle-offset-git.patch](https://github.com/user-attachments/files/27954953/jellyfin-androidtv-subtitle-offset-git.patch)

before applying the patch I did checkout to the latest release which is 0.19.9 but unfortunately things escalated quickly:

`
...
error: app/src/main/java/org/jellyfin/androidtv/ui/playback/VideoManager.java: patch does not apply
error: patch failed: app/src/main/java/org/jellyfin/androidtv/ui/player/video/VideoPlayerControls.kt:44
error: app/src/main/java/org/jellyfin/androidtv/ui/player/video/VideoPlayerControls.kt: patch does not apply
error: patch failed: app/src/main/res/values-en-rGB/strings.xml:646
error: app/src/main/res/values-en-rGB/strings.xml: patch does not apply
error: patch failed: app/src/main/res/values/strings.xml:660
error: app/src/main/res/values/strings.xml: patch does not apply
error: patch failed: playback/core/src/main/kotlin/backend/PlayerBackend.kt:40
error: playback/core/src/main/kotlin/backend/PlayerBackend.kt: patch does not apply
error: patch failed: playback/core/src/main/kotlin/backend/PlayerBackendEventListener.kt:7
...
`
so, which branch should be the base when using this patch?

edit: ok, I'm on master now and no errors after applying the patch :)  
edit: well done sir, that is a very well done job

**Comment by ErikDB87** [comment]
@pulpul-s, will you submit this as a PR?

I obviously feel you should do this, because you deserve all the credit. But if you don't want to to (for whatever reason), I will give that a try (obviously citing you as the author). I just think the feature is really missed, and should land in an official release asap. And the first step is of course getting it into `master`.

(I added the APK based on `master` as a "release" in [my forked repository](https://github.com/ErikDB87/jellyfin-androidtv). - Apologies if that link is also not allowed...)

**Comment by pulpul-s** [comment]
> [@pulpul-s](https://github.com/pulpul-s), will you submit this as a PR?
> 
> I obviously feel you should do this, because you deserve all the credit. But if you don't want to to (for whatever reason), I will give that a try (obviously citing you as the author). I just think the feature is really missed, and should land in an official release asap. And the first step is of course getting it into `master`.
> 
> (I added the APK based on `master` as a "release" in [my forked repository](https://github.com/ErikDB87/jellyfin-androidtv). - Apologies if that link is also not allowed...)

You can do a PR if you feel that the code is ready for it. From my part, I feel it is not clean enough for a PR, because it is something I just stitched together out of frustration, to be honest I can't even remember what I did. And unfortunately like I said earlier I'm stuck with other projects so I really don't have time to attend to a PR. Sorry about that.

You are free to use and improve it in any ways you feel is necessary. Maybe ask an AI agent for a review and test it thoroughly.

**Comment by ErikDB87** [comment]
> Maybe ask an AI agent for an review and test it thoroughly

I'll start with that. My coding knowledge is too small to see any issues or possible improvements myself. 

And after that, we'll see what the comments are 😉

**Comment by dev-Blaze** [comment]
any updates on this?

**Comment by camarigor** [comment]
Hi. I run a small Jellyfin server for a few people. Some of them watch on a Fire TV Stick and they keep asking me why they cannot nudge the subtitles when the timing is off. I went looking and found this issue, and then #5597.

I know #5597 was closed for the LLM policy and not for technical reasons. I read the policy. I am not trying to bring that diff back, I would write this myself.

Before asking anything here I spent some time in the media3 source, version 1.11.0, which is the one pinned in gradle/libs.versions.toml:20. To be clear about scope, I only mean subtitle timing. Audio delay is a different problem and I am not proposing to solve it.

What I found is that there are three ways to do this and each one costs something. That is where I got stuck.

The first is what #5597 did. About 220 lines in seven small classes, and the offset can be changed while playing. But it depends on TextRenderer.experimentalSetLegacyDecodingEnabled, and in 1.11.0 that method is deprecated. The javadoc says legacy subtitle decoding will be removed in a future release (TextRenderer.java:294).

The second is writing our own renderer for cues, which avoids the deprecated API. Here TextRenderer is final (TextRenderer.java:71) so it cannot be extended, and RendererHolder checks renderer instanceof TextRenderer in two places, line 236 for setFinalStreamEndPositionUs and line 401 for the seamless transition. A renderer that only delegates never gets the stream end, so the inner TextRenderer stays on C.TIME_UNSET. CuesResolver and its implementations are package private on top of that, so cue resolution would have to be written again and rechecked on every media3 bump. This is the expensive one and I am not sure it is worth it.

The third is applying the offset in SubtitleParser, during extraction. Much smaller and only supported API. The catch is the cues are already parsed and buffered, so changing the offset needs the source prepared again.

If it was up to me alone I would take the third one and live with that. But you wrote in #5692 that subtitle and audio track selection come first for the new player, and the legacy player is still the default (UserPreferences.kt:155), so if the answer is not now, that is fine.

So my question is simple. Would you accept any of these three, or is there a fourth one I am not seeing? I will not start writing code before you answer, and if you do not answer I will read that as a no.

**Comment by nielsvanvelzen** [comment]
We explicitly don't want to re-implement extractors, renderers and other media3 implementations within our app. As that would introduce a huge maintenance burden for us. So I think that changes would need to be made upstream to expose this functionality before we can implement it.

**Comment by camarigor** [comment]
Thanks, that is a clear answer and it lines up with what I found, so I will not open a PR here.

For anyone landing on this thread later, the upstream request is https://github.com/androidx/media/issues/1976, open since December 2024. I have posted the specific blockers there: the three ways an app could do this today are all closed off (`TextRenderer` is final, `RendererHolder` branches on the concrete type in two places, `CuesResolver` and its implementations are package private) and the one remaining hook is deprecated with removal already announced.

So this is genuinely blocked on media3 rather than on someone writing the code here.

**Comment by ErikDB87** [comment]
I see @nielsvanvelzen responded with a 'hoorey' emoji to https://github.com/androidx/media/issues/1976#issuecomment-5437589632, stating that https://github.com/androidx/media/pull/3370 was merged. 

Hopefully this means a step in the right direction, @camarigor? :)

**Comment by nielsvanvelzen** [comment]
Correct, we should be able to (finally) implement the subtitle offset feature once the next version of media3 releases!

**Comment by camarigor** [comment]
That is genuinely good news, and full credit to @fluffypony for landing it. The underlying request goes back to google/ExoPlayer#854 in 2015, so this has been a long wait.

I had a look at where it leaves this app, and it is better than I expected. `VideoManager` already builds a `MediaItem.SubtitleConfiguration` for every subtitle stream Jellyfin delivers as `EXTERNAL`, which means the new `timeOffsetUs` field applies to objects the app is already constructing today. When I first read this issue I assumed any fix would need new plumbing, and for external subtitles it looks like the plumbing is already there. Embedded and HLS tracks are not sideloaded, so they sit outside this one.

The other part I like: the offset is meant to change mid playback through `replaceMediaItem` without re-buffering. That is the difference between a real sync control and something that restarts the video every time you nudge it.

It is not published yet, the merge is on `main` and 1.11.0 predates it, so this is next release material. Happy to help however is useful when it gets there.
